Use journalling to mark deletion of objects in remote tier

krisis opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

krisis commented:
@poornas this is still WIP. The following issues remain to be fixed,
- need to handle permanently offline 'first' disk
- deleteTransitionedObjects throws the following error, need to investigate.
```
 Error: Operation timed out (cmd.OperationTimedOut)
minio1_1  |        3: cmd/bucket-lifecycle.go:231:cmd.deleteTransitionedObject()
minio1_1  |        2: cmd/bucket-handlers.go:618:cmd.objectAPIHandlers.DeleteMultipleObjectsHandler()
minio1_1  |        1: net/http/server.go:2042:http.HandlerFunc.ServeHTTP(...
```

poornas commented on cmd/endpoint.go:
```diff
@@ -256,6 +256,20 @@ func (l EndpointServerPools) Localhost() string {
+		}
+	}
+	return "", false
+}
```
we will need to check if this disk is online rather than assume it is

poornas commented on cmd/bucket-lifecycle.go:
```diff
@@ -200,44 +200,55 @@ func validateTransitionDestination(sc string) error {
+	expireRestoredObj
+	// expireRemoteObjOnly indicates deletion of transitioned objects from the remote tier.
+	expireRemoteObjOnly
+)
```
much nicer!

poornas commented on cmd/bucket-handlers.go:
```diff
@@ -481,6 +481,24 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
+					VersionID: object.VersionID,
+				}
+				continue
+			}
```
this code block is unnecessary. To do a GetObjectInfo you will just need to pass VersionID in the ObjectOptions
```
opts := ObjectOptions{VersionID: object.VersionID}
if opts.VersionSuspended && opts.VersionID == "" {
    opts.VersionID = nullVersionID
}
goi, gerr = getObjectInfoFn(ctx, bucket, object.ObjectName, opts)
```

poornas commented on cmd/object-handlers.go:
```diff
@@ -1617,6 +1617,59 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
+		}
+		if pOpts.VersionSuspended && pOpts.VersionID == "" {
+			pOpts.VersionID = nullVersionID
+		}
```
putOpts is not required for GetObjectInfo call
you can directly check globalBucketVersioningSys.Suspended(bucket) and use opts.VersionID (opts already initialized ahead)

krisis committed:
Use journalling to mark deletion of objects in remote tier

- Also simplify helper function to cleanup transitioned objects on expiry or when
they are 'overwritten'
- Use expireTransitionedObjects in delete-object, multi-delete-objects and
put-object API handlers

poornas commented on cmd/object-handlers.go:
```diff
@@ -1618,6 +1655,27 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
+		// Get appropriate object info to identify the remote object to delete
+		goiOpts = ObjectOptions{
+			VersionID: opts.VersionID,
+		}
```
```suggestion
		goiOpts = ObjectOptions{}
```
versionID is generated by MinIO, not specified in request headers unless it is replication.

poornas commented on cmd/server-main.go:
```diff
@@ -532,6 +532,10 @@ func serverMain(ctx *cli.Context) {
 		}
+		globalTierJournal, err = initTierDeletionJournal(GlobalContext.Done())
+		if err != nil {
+			logger.FatalIf(err, "Unable to initialize remote tier pending deletes journal ")
```
```suggestion
			logger.FatalIf(err, "Unable to initialize remote tier pending deletes journal")
```

poornas commented on cmd/tier-journal.go:
```diff
@@ -0,0 +1,278 @@
+	w, err := globalTierConfigMgr.getDriver(tierName)
+	if err != nil {
+		return err
+	}
```
if tier name is not found, seems to me we should just drop the entry rather than recirculating it in the journal forever. thoughts?

poornas commented on cmd/data-scanner.go:
```diff
@@ -958,11 +958,12 @@ func applyTransitionAction(ctx context.Context, action lifecycle.Action, objLaye
 			Versioned: globalBucketVersioningSys.Enabled(obj.Bucket),
 			MTime:     obj.ModTime,
-		}); err != nil {
+		}); err != nil && err != toObjectErr(errMethodNotAllowed, obj.Bucket, obj.Name) {
```
why even attempt a Copy when it is a delete marker. 
```
 if obj.DeleteMarker{
   return false
}
```
On second thoughts, if it is a delete marker - it should not be returning TransitionAction. If this is because of the debug variable, we should avoid creating a TransitionAction there for a delete marker instead of this type of check.

krisis replied:
@poornas this bug exists in `pkg/bucket/lifecycle/lifecycle.go` even today. I shall send a separate PR to fix this.

poornas commented on cmd/web-handlers.go:
```diff
@@ -808,14 +812,12 @@ next:
+				if jentry, ok := os.ShouldRemoveRemoteObject(); ok {
+					err := globalTierJournal.AddEntry(jentry)
+					logger.LogIf(ctx, err)
+				}
```
Upload overwrite should also be handled for web handlers.

poornas commented on cmd/web-handlers.go:
```diff
@@ -754,8 +753,13 @@ next:
 				hasLifecycleConfig = true
 			}
+			os := newObjSweeper(args.BucketName, objectName)
+			opts = os.GetOpts()
```
if version is suspended, overwriting the empty version with "null" would cause a permanent delete of that version as opposed to setting a delete marker - same bug in all the other delete calls

poornas commented on cmd/object-handlers.go:
```diff
@@ -2804,10 +2870,21 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
+		// affect its null version. Through opts below we select
+		// the null version's remote object to delete if
+		// transitioned.
+		goiOpts = os.GetOpts()
```
if versioning is suspended on the bucket, null version of object (which may not be at the top of the stack) will be permanently deleted instead of setting a delete marker.

krisis replied:
As discussed over the call, null version is handled correctly. I have added a comment to objSweeper.ReqVersion to clarify its origin.

poornas commented on cmd/object-handlers.go:
```diff
@@ -2913,26 +2990,11 @@ func (api objectAPIHandlers) DeleteObjectHandler(w http.ResponseWriter, r *http.
-			}, goi.transitionedObjName, goi.TransitionTier, false, false)
+	if hasLifecycleConfig {
+		// Remove the transitioned object whose object version is being overwritten.
+		if jentry, ok := os.ShouldRemoveRemoteObject(); ok {
```
ShouldRemoveRemoteObject should not automatically delete the remote object if versioning is suspended on bucket - I think this attempt to unify can cause way more bugs if we are not careful

poornas commented on cmd/bucket-handlers.go:
```diff
@@ -606,23 +616,18 @@ func (api objectAPIHandlers) DeleteMultipleObjectsHandler(w http.ResponseWriter,
+
+	// Clean up transitioned objects from remote tier
+	if hasLifecycleConfig {
+		for _, os := range oss {
```
If there are errors due to permissions or object locking constraints, the oss can have null entries - need to guard against panics

poornas reviewed (approved):
LGTM
